src/net/cruhland/Analysis/Chapter4/Integers.agda

```diff
@@ -238,10 +238,30 @@ _ = AA.zero-prod {{r = ℤ.zero-product}}
 -- Corollary 4.1.9 (Cancellation law for integers). If a, b, c are
 -- integers such that ac = bc and c is non-zero, then a = b.
 -- Exercise 4.1.6
-_ : {a b c : ℤ} → c ≄ 0 → a * c ≃ b * c → a ≃ b
-_ = λ c≄0 →
-      let instance c≄ⁱ0 = fromWitnessFalse c≄0
-       in AA.cancel {{r = ℤ.*-cancellativeᴿ}}
+record IntegerCancellation : Set where
+  field
+    proof : {a b c : ℤ} → c ≄ 0 → a * c ≃ b * c → a ≃ b
+
+_ : IntegerCancellation
+_ = record { proof = λ c≄0 →
+                   let instance c≄ⁱ0 = fromWitnessFalse c≄0
+                   in AA.cancel {{r = ℤ.*-cancellativeᴿ}}
+           }
+
+CancellationConverse : IntegerCancellation → {a b : ℤ} → a * b ≃ 0 → a ≃ 0 ∨ b ≃ 0
```

Comment:
I will add some comments here, that we are proving that Integercancellation implies ZeroDivisor, which is converse of proof for Exercise 4.1.6 above.